torch/csrc/jit/codegen/cuda/index_compute.cpp

```diff
@@ -1939,6 +1887,241 @@ std::pair<std::vector<kir::Val*>, bool> Index::getConsumerRootPredIndices(
+  auto root_ids = ir_utils::filterByType<IterDomain>(root_vals);
+  std::vector<IterDomain*> contiguous_ids(root_ids.begin(), root_ids.end());
+
+  if (root_vals.empty()) {
+    return std::vector<PredicateContigInfo>();
+  }
+
+  auto exprs = ExprSort::getExprs(
+      (*root_vals.begin())->fusion(),
+      {reference_domain.begin(), reference_domain.end()});
+
+  for (auto expr : exprs) {
+    if (expr->isA<Merge>()) {
+      auto merge = expr->as<Merge>();
+      auto inner_contig_it = std::find(
+          contiguous_ids.begin(), contiguous_ids.end(), merge->inner());
+      auto outer_contig_it = std::find(
+          contiguous_ids.begin(), contiguous_ids.end(), merge->outer());
+      if (inner_contig_it != contiguous_ids.end() &&
+          outer_contig_it != contiguous_ids.end()) {
+        contiguous_ids.erase(outer_contig_it);
+        // iterators invalid now, do a find
+        contiguous_ids.erase(std::find(
+            contiguous_ids.begin(), contiguous_ids.end(), merge->inner()));
+        contiguous_ids.emplace_back(merge->out());
+      }
+    }
+  }
+
+  std::vector<PredicateContigInfo> contig_id_infos;
+
+  // Contig id is based on reference, need to transform to a concrete value.
+  const auto gpu_lower = GpuLower::current();
+
+  for (auto contig_id : contiguous_ids) {
+    auto contig_root_vals = IterVisitor::getInputsTo({contig_id});
+    auto contig_root_ids = ir_utils::filterByType<IterDomain>(contig_root_vals);
+    PredicateContigInfo contig_id_info;
+    contig_id_info.contig_id = contig_id;
+    std::transform(
+        contig_root_ids.begin(),
+        contig_root_ids.end(),
+        std::inserter(contig_id_info.root_ids, contig_id_info.root_ids.begin()),
+        [&gpu_lower](IterDomain* root_id) {
+          return gpu_lower->caLoopMap().getConcreteMappedID(root_id);
+        });
+    contig_id_infos.push_back(contig_id_info);
+  }
+  return contig_id_infos;
+}
+
+} // namespace
+
+std::pair<std::vector<kir::Bool*>, std::vector<std::unordered_set<IterDomain*>>>
+Index::getReferenceRootPredicates(
+    const kir::TensorView* kir_consumer_tv,
+    const std::vector<kir::ForLoop*>& loops,
+    bool unswitch) {
+  FUSER_PERF_SCOPE("Index::getConsumerRootPredIndices");
+
+  const auto gpu_lower = GpuLower::current();
+  kir::IrBuilder ir_builder(gpu_lower->kernel());
+
+  // Get a reference tensor replayed as existing loop structure
+  ReferenceTensor reference = IndexReferenceReplay::getReference(loops);
+  auto reference_domain = reference.domain;
+  auto reference_id_map = reference.concrete_to_id;
+
+  std::unordered_map<kir::ForLoop*, kir::Val*> loop_to_ind_map;
+
+  std::transform(
+      loops.begin(),
+      loops.end(),
+      std::inserter(loop_to_ind_map, loop_to_ind_map.begin()),
+      [&ir_builder](kir::ForLoop* fl) {
+        if (fl->iter_domain()->isBroadcast()) {
+          return std::make_pair(fl, ir_builder.zeroVal()->as<kir::Val>());
+        }
```

Comment:
Isn't it guaranteed that the index of a broadcast loop is `zeroVal`? Do we really need to have this branch?

Reply:
You're right, it should be, but for some reason seemed not to be the case. I'll look into it more.

Reply:
test_data_compatability in our python tests has issues with this:
```
FOR 0 in id35[blockIdx.x.sb(0 .. i51)]:
  IF Unswitch:
    FOR 0 in id36[US.sb(0 .. 1)] UNROLL:
      FOR 0 in id34[S.sb(0 .. 1)] UNROLL:
        FOR 0 in id32[threadIdx.x.sb(0 .. 128)] UNROLL:
          T3 = __half2float(T2)
```

Reply:
Sorry, still looking, the issue was with the below unswitch :
```
      if (within_unswitch) {
        if (loop->iter_domain()->isBroadcast()) {
          continue;
        }
```

Reply:
okay, yes this was safe to remove.